Makefile

```diff
@@ -58,8 +58,15 @@ checksum:
 clean:
 	@rm -rf $(BINDIR) ./_dist
 
+GIT_BASEDIR    = $(shell git rev-parse --show-toplevel 2>/dev/null)
+ifneq ($(GIT_BASEDIR),)
+	LDFLAGS += -X github.com/Azure/acs-engine/pkg/test.JUnitOutDir=${GIT_BASEDIR}/test/junit
+endif
+
 test: test-style
-	go test -v $(GOFILES)
+	go get github.com/onsi/ginkgo/ginkgo
+	go get github.com/onsi/gomega
```

Comment:
My thinking is that the Docker Hub microsoft/acs-engine image IS the development environment image anyone can use or we just use ourselves.  We could automate the push of the images to Docker hub easily using CircleCI and tag the image with the short git commit when it hits master - then you can match any image in Docker hub to any commit in Github off master (your point 5).   The idea is that the public docker hub image makes it easier to get the development environment for anyone versus building the docker image every time locally.  If we don't want them to have to update every commit to master, we could be more selective of when we push to docker hub (i.e. per release or when a significant change happens with Dockerfile).

Per point 1 - technically we don't have to build an image during CI (which I know we're doing now) to do the e2e tests.  It is really a safeguard measure for if/when someone upgrades the depdendent components (kubectl, az cli, etc).  If we would extract these type of components from the build image, then I think this would be great as they aren't really needed to build acs-engine binary itself or generate templates, right?  This would also help with your point #2 and 3.

In general, I'm all about lightweight docker image which just has binaries to simply use acs-engine (versus download the binary).  This perhaps would be best in a separate image.